### ISSUE TITLE
False-positives where templated methods accepts closures

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#826]
https://getpsalm.org/r/fa83e8b8fc
```php
<?php
/** 
 * @template TKey
 * @template TValue
 */
class Collection {
    /**
     * @param Closure(TValue):bool $p
     * @return Collection<TKey,TValue>
     */
    public function filter(Closure $p);
}
class I {}

/** @var Collection<mixed,Collection<mixed,I>> $c */
$c = new Collection;

$c->filter(
  /** @param Collection<mixed,I> $elt */
  function(Collection $elt): bool { return (bool) rand(0,1); }
);

$c->filter(
  /** @param Collection<mixed,I> $elt */
  function(Collection $elt): bool { return true; }
);
```
**Expected**: no issues
**Actual**: the first call to `filter()` complains that it needs `Closure(Collection<mixed,I>):bool` but it got `Closure(Collection<mixed,I>):bool` instead (I don't see any difference in those types though). The second one doesn't like closure returning `true` where it expects closure returning `bool`.

**Comment by muglug** [comment]
It's a bug with https://github.com/vimeo/psalm/blob/2be0f68d874eff5f00c420c9ca28a9eeb7baceb5/src/Psalm/Type/Union.php#L748 - this should take union types, not strings - was an unnecessary optimisation
